libdevice/cmath_wrapper_fp64.cpp

```diff
@@ -131,3 +131,274 @@ double asinh(double x) { return __devicelib_asinh(x); }
+        ps->_Sh[_D2] = (unsigned short)(ps->_Sh[_D2] >> xexp |
+                                        ps->_Sh[_D1] << (16 - xexp));
+        ps->_Sh[_D1] = (unsigned short)(ps->_Sh[_D1] >> xexp |
+                                        ps->_Sh[_D0] << (16 - xexp));
+        ps->_Sh[_D0] >>= xexp;
+      }
+
+      ps->_Sh[_D0] |= sign;
+      if ((0x8000 < psx || 0x8000 == psx && (ps->_Sh[_D3] & 0x0001) != 0) &&
+          (++ps->_Sh[_D3] & 0xffff) == 0 && (++ps->_Sh[_D2] & 0xffff) == 0 &&
+          (++ps->_Sh[_D1] & 0xffff) == 0) {
+        ++ps->_Sh[_D0]; // round up
+      } else if (ps->_Sh[_D0] == sign && ps->_Sh[_D1] == 0 &&
+                 ps->_Sh[_D2] == 0 && ps->_Sh[_D3] == 0) {
+        return 0;
+      }
+      return _FINITE;
+    }
+  }
+}
+
+DEVICE_EXTERN_C
+short _Exp(double *px, double y,
+           short eoff) { // compute y * e^(*px), (*px) finite, |y| not huge
+  static const double invln2 = 1.4426950408889634073599246810018921;
+  static const double c1 = 22713.0 / 32768.0;
+  static const double c2 = 1.4286068203094172321214581765680755e-6;
+  static const double p[] = {1.0, 420.30235984910635, 15132.70094680474802};
+  static const double q[] = {30.01511290683317, 3362.72154416553028,
+                             30265.40189360949691};
+
+  _Dconst _Eps = {INIT((_DBIAS - NBITS - 1) << _DOFF)};
+  _Dconst _Inf = {INIT(_DMAX << _DOFF)};
+  if (*px < -HUGE_EXP || y == 0.0) { // certain underflow
+    *px = 0.0;
+    return 0;
+  } else if (HUGE_EXP < *px) { // certain overflow
+    *px = _Inf._Double;
+    return _INFCODE;
+  } else { // xexp won't overflow
+    double g = *px * invln2;
+    short xexp = (short)(g + (g < 0.0 ? -0.5 : +0.5));
+    g = xexp;
+    g = (*px - g * c1) - g * c2;
+    if (-_Eps._Double < g && g < _Eps._Double) {
+      *px = y;
+    } else { // g * g worth computing
+      const double z = g * g;
+      const double w = (q[0] * z + q[1]) * z + q[2];
+
+      g *= (z + p[1]) * z + p[2];
+      *px = (w + g) / (w - g) * 2.0 * y;
+      --xexp;
+    }
+    return _Dscale(px, (long)xexp + eoff);
+  }
+}
+
+DEVICE_EXTERN_C
+double _Cosh(double x, double y) { // compute y * cosh(x), |y| <= 1
+  switch (_Dtest(&x)) {            // test for special codes
+  case _NANCODE:
+  case _INFCODE:
+    return x;
+  case 0:
+    return y;
+  default: // finite
+    if (y == 0.0) {
+      return y;
+    }
+
+    if (x < 0.0) {
+      x = -x;
+    }
+
+    if (x < _Xbig) { // worth adding in exp(-x)
+      _Exp(&x, 1.0, -1);
+      return y * (x + 0.25 / x);
+    }
+    _Exp(&x, y, -1);
+    return x;
+  }
+}
+
+DEVICE_EXTERN_C
+double _Poly(double x, const double *tab, int n) { // compute polynomial
+  double y;
+
+  for (y = *tab; 0 <= --n;) {
+    y = y * x + *++tab;
+  }
+
+  return y;
+}
+
+DEVICE_EXTERN_C
+double _Sinh(double x, double y) { // compute y * sinh(x), |y| <= 1
+
+  short neg;
+  // coefficients
+  static const double p[] = {0.0000000001632881, 0.0000000250483893,
```

Comment:
How did you get these?

Reply:
This patch is based on Microsoft STL(https://github.com/microsoft/STL). All the algorithm and magic number are from Microsoft's implementation.

Reply:
As I understand, when we do not have the 64-bit FP limitation, we would just include MSVC STL header files, and get rid of this re-definitions.

Reply:
Hi, @vzakhari 
In our cases, including Microsoft STL headers can't solve the problem. Our internal msvc functions included in this patch is in library but not header files. If we search in Visual studio package, we can find the implementation but they are in ".c" files instead of ".h" files.
MSVC std::complex's implementation are in <complex> headers, if we use std::complex in sycl kernel, we will encounter compiling error with the message "undefined reference _FSinh..." which means compiler can't find the implementation in headers, the undefined references are left to be resolved by linker using some MSVC runtime library.

Reply:
Hi @jinge90, I was actually referring to the macro and type definitions that you had to duplicate.  If/when we can include msvc header files, the code will be a little bit cleaner.

Reply:
Hi, @vzakhari 
We have considered to remove all macro definition by including MSVC headers. The patch used some macro, some of them are from standard headers such as <math.h>, we can just include <math.h>, these macros include _FSIGN, _FFRAC, _FMASK....
However, some macros are defined in MSVC internal headers which are not exposed to developers. For exmple, _F0, _F1, NBITS..., are defined in "xmath.h", we can't "#include <xmath.h>" in source code, so we have to define in source code.
Thank you very much.

Reply:
Oh, I see.  These macros and types are not exposed to the user, so we are doomed to duplicate them :)